lib/ansible/galaxy/dependency_resolution/providers.py

```diff
@@ -126,9 +129,11 @@ def identify(self, requirement_or_candidate):
 
     def get_preference(
             self,  # type: CollectionDependencyProvider
-            resolution,  # type: Optional[Candidate]
-            candidates,  # type: List[Candidate]
-            information,  # type: List[NamedTuple]
+            identifier,  # type: str
+            resolutions,  # type: Mapping[str, Candidate]
+            candidates,  # type: Mapping[str, Iterator[Candidate]]
+            information,  # type: Mapping[str, Iterable[PreferenceInformation]]
```

Comment:
This is a breaking change that will make the downstream packaging hard. The API compatibility should be with all the minor versions stated in the previous review.

Reply:
resolvelib is a relatively new library, which not much distribution are now having official package for it:
* https://pkgs.org/download/python3-resolvelib (only Ubuntu 21.04+, Debian 11+, openSUSE Leap 15.2+, Fedora 34+)

Moreover, if ansible 2.12.0 already drop Python support <3.7, I could see most distributions (e.g. CentOS 7, CentOS 8, Debian 10, openSUSE Leap 15.2, openSUSE Leap 15.3, Ubuntu 18.04, etc) are no longer supported.

In this case, the most workable installation method would be:
* Traditional `pip install ansible` or `pipx install ansible` which goes to `/usr/local/lib` (therefore, no downstream packaging difficulties), or
* Someone crazily packaging all related package with latest compatible version deb/rpm (which go for something like my https://build.opensuse.org/project/show/home:alvistack solution), or
* Modern upcoming distribution bundling Python 3.8+, resolvelib 0.8.1+, ansible 2.12.0+ all together (in the future release)

For either 3 cases we have no strong reason to pin resolvelib <0.6.0, isn't it?

Reply:
> Moreover, if ansible 2.12.0 already drop Python support <3.7, I could see most distributions (e.g. CentOS 7, CentOS 8, Debian 10, openSUSE Leap 15.2, openSUSE Leap 15.3, Ubuntu 18.04, etc) are no longer supported.

That would be an incorrect assumption.  While those distros may have a default python lower than 3.8, many of them provide newer python versions that can still be installed.

Reply:
You missed our main downstream which is RHEL. It's usually non-trivial to get packages updated there. Plus I'm not sure if Ubuntu PPAs have caught up. And what about the rolling-release distros like Gentoo/Arch where some software may still need an older version, for example?

Ansible is not pip, we don't vendor resolvelib. This is why we cannot just go and drop the lower bounds. Historically, we've always kept high compatibility with older systems. I've requested more reviews to gather opinions.

Reply:
I am in agreement with @webknjaz 

Please keep the lower bounds of the original requirements.

Reply:
> That would be an incorrect assumption. While those distros may have a default python lower than 3.8, many of them provide newer python versions that can still be installed.

* CentOS 7: SCL? https://centos.pkgs.org/7/centos-sclo-rh-x86_64/rh-python38-2.0-4.el7.x86_64.rpm.html
* CentOS 8: AppStream? https://centos.pkgs.org/8-stream/centos-appstream-x86_64/python38-3.8.12-1.module_el8.6.0+929+89303463.x86_64.rpm.html
* Ubuntu 18.04: PPA? https://launchpad.net/~deadsnakes/+archive/ubuntu/ppa
* Debian 10: Personal repo? https://people.debian.org/~paravoid/python-all/

For neither above cases, you couldn't find resolvelib prepackaged (see https://pkgs.org/download/python3-resolvelib) for ansible 2.12.0+

If we could not even find a resolvelib <0.6.0 official deb/rpm with above legacy 2.5+ years ago distribution, we need someone package it as deb/rpm, or people need to install it with pip/pipx.

Reply:
> You missed our main downstream which is RHEL. It's usually non-trivial to get packages updated there. Plus I'm not sure if Ubuntu PPAs have caught up. And what about the rolling-release distros like Gentoo/Arch where some software may still need an older version, for example?

When ansible 2.12.0+ choose to drop Python <3.7 support, our main downstream RHEL 7/8 could only pin with ansible 2.11.6+ (unless these downstream also packaging both Python 3.8 + all python dependencies into an official supported repo); for RHEL 9 beta it is Python 3.9 which shouldn't have difficulties for both ansible 2.12.0+ and resolvelib 0.8.1+?